Implements Zero-Copy String Views onto Contiguous Memory.

daniel-j-h opened the pull request:
For https://github.com/Project-OSRM/osrm-backend/issues/3265. Work in progress. This is harder than I thought, mostly because call sites often mutate the string which would mean we have to make a copy.

Also cleans up some type aliases on the way.

## Tasklist
 - [x] re-write call sites using string views
 - [x] rip out old-style string interface
 - [x] review
 - [x] adjust for comments

References:

- http://www.boost.org/doc/libs/1_61_0/libs/utility/doc/html/string_ref.html
- http://en.cppreference.com/w/cpp/string/basic_string_view

daniel-j-h commented on include/engine/datafacade/contiguous_internalmem_datafacade_base.hpp:
```diff
@@ -729,7 +729,30 @@ class ContiguousInternalMemoryDataFacadeBase : public BaseDataFacade
+        auto last = m_names_char_list.begin() + range.back() + 1;
+        // These iterators are useless: they're InputIterators onto a contiguous block of memory.
+        // Deref to get to the first element, then Addressof to get the memory address of the it.
+        auto len = &*last - &*first;
```
Ticketed in https://github.com/Project-OSRM/osrm-backend/issues/3458

daniel-j-h commented on include/util/integer_range.hpp:
```diff
@@ -39,6 +39,7 @@ template <typename Integer> class range
 
     // Iterator functions
     bool operator!=(const range &) const noexcept { return iter < last; }
+    bool operator==(const range &other) const noexcept { return !operator!=(other); }
```
Ticketed a long time ago: https://github.com/Project-OSRM/osrm-backend/issues/2859 this is biting us down the road..

daniel-j-h commented:
Status: call sites needs to be adapted. Some of them mutate the strings, some of them make comparisons with / lookups against sets of strings - this all needs to be re-written. Can be done incrementally now.

TheMarex commented:
> Some of them mutate the strings

Where would that be? Can you ticket those location, that could be code smell.

daniel-j-h commented:
For example

https://github.com/Project-OSRM/osrm-backend/blob/b1f6797aab453e141931af2389b8e16687e47fd6/include/util/guidance/name_announcements.hpp#L29-L40

lower case prefix and suffix, creating new strings out of the strings we get from the underlying storage.

daniel-j-h commented:
I had some time to revisit this - implemented all string view features, there's one call site left which needs to be adapted: the [mutations here](https://github.com/Project-OSRM/osrm-backend/blob/521492b3eb696e7284486a7f9d42e480ad3ce6ec/include/util/guidance/name_announcements.hpp#L36-L37) and up the call chain. The logic needs to be changed to work on immutable string views only. Squashed and rebased onto master.

TheMarex commented:
@daniel-j-h could we fix this for the moment by doing a copy to a string inside the function? If the algorithm is adjusted to work on read-only data we can remove this without changing the callsite.

daniel-j-h commented:
@TheMarex rebased against master and updated to recent changes. I ripped out all old string datafacade functions now for good. Two issues I'd like to flag here:
- mutation in the suffix function, making a copy of the strings
- in order to fully benefit from zero-copy we should expose string views to the outside (api)

TheMarex commented on include/engine/datafacade/contiguous_internalmem_datafacade.hpp:
```diff
@@ -856,19 +859,19 @@ class ContiguousInternalMemoryDataFacade : public BaseDataFacade
+                    m_datasource_name_lengths[id];
+        // These iterators are useless: they're InputIterators onto a contiguous block of memory.
+        // Deref to get to the first element, then Addressof to get the memory address of the it.
+        const std::size_t len = &*last - &*first;
```
Why not use `std::distance`?

daniel-j-h replied:
Because these are pointers. Always.
Also `operator-` is guaranteed to work for random access iterators anyway.

TheMarex commented on include/util/string_view.hpp:
```diff
@@ -0,0 +1,34 @@
+{
+template <> struct hash<::osrm::util::StringView> final
+{
+    std::size_t operator()(::osrm::util::StringView v) const noexcept
```
Why is `v` taken by value here? Is there a downside to `const&`?

daniel-j-h replied:
`StringView` is basically `struct StringView { const char* first; const char* last; }` - better to pass by value. The `StringView` is immutable by default, too, so there is no real benefit of passing it by `const`.

(If you're asking yourself, why immutable by default, consider `StringView s{"hello"}; s[0] = 'w'; // oops`)

TheMarex replied:
Thanks for the clarification. 👍

TheMarex commented on src/extractor/suffix_table.cpp:
```diff
@@ -11,14 +13,23 @@ namespace extractor
+    return isSuffix(util::StringView{possible_suffix});
+}
+
+bool SuffixTable::isSuffix(util::StringView possible_suffix) const
```
Same here: StringView passed by value.

TheMarex commented on include/engine/datafacade/contiguous_internalmem_datafacade.hpp:
```diff
@@ -716,56 +716,59 @@ class ContiguousInternalMemoryDataFacade : public BaseDataFacade
     }
 
-    std::string GetNameForID(const unsigned name_id) const override final
+    virtual StringView GetNameForID(const NameID id) const override final
```
Most functions here omit the `virtual` since it is not necessary.

daniel-j-h replied:
Yeah will adapt it to be consistent with the rest
